### PR TITLE
Integrate pyroengine latest changes

### PR DESCRIPTION
**Pull request by chloeskt**
Hi everyone ! 🤗 

This is an important PR, I am sorry for its length. Here I will explain the general purpose of it and I added a review with comments on certain part of the code to guide you. 

### Motivations & changes

- Integrate recent changes in pyroengine such as the introduction of the webserver and the monitor_pi.py script. This integration has demanded some new playbooks to be introduced.
- Make it so that Ansible can be run on the master rpi which will use it to configure the rpi with camera (in particular copy some files from pyroengine) and later, update itself and them (scope of another PR).
- Add some configuration variables to ensure that Ansible has more time and persistant connection to retry reaching hosts (sometimes the ssh connection has a bit of a lag making your local machine unable to reach the rpis). These variables are supposed to help to a certain extent. 
- Reshape the architecture of the repo to follow conventions on templates and files directories notably and make it cleaner.
- Update the READMEs consequently.
- For the playbooks running on the main rpi (i.e. the installed on it and run by the Ansible installed on it - cf. `files/ansible_on_main_rpi/playbooks`), add the use of this [module](https://docs.ansible.com/ansible/latest/collections/community/general/slack_callback.html#parameters) which sends msg while the playbook is running. To do so, we need to specify as an environment variable `SLACK_WEBHOOK_URL` (Incoming Webhook app on Slack). Careful this variable MUST NOT be committed, you need to add it in `vars/main.yml`. The file `vars/main.yml.dist` serves as template. On your Slack I've already deployed this app, send me a DM in you want the url. Below you have two screenshots of what is send (you can see the name of the host on which the playbook is run, which palybook is run and the exit status). This will allow us, once updates are send to check whether or not the updates on the rpis were successful. 
- In `files/ansible_on_main_rpi/templates` there is a template file `python.env.dist` which is meant to help you create a file `python.env` that contains the fixed IP address of the main rpi on which the local webserver is run and the corresponding port. This file will be sent by the main rpi to the rpi with camera so that they know the url to use when doing a post request (use the local webserver).

Again sorry for the length of this PR, it is now difficult to break it down. 😅 \
Happy to have your feedback and discuss it ! 

#### When a playbook has failed
Capture d’écran 2021-04-11 à 22 46 46

#### When a playbook has succeeded
Capture d’écran 2021-04-11 à 23 08 17

### CODE REVIEW

File: main.rpi.requirements.txt
Comment (fe51) [inline main.rpi.requirements.txt]: This is more about naming of the file. 
Either the host is raspberrypi_master or raspberrypi_main, but to it would be better to keep the same term everywhere to avoid confusion. What do you think ?

File: README.md
Comment (chloeskt) [inline README.md:80]: In this PR, we allowed Ansible to be installed on the main rpi, to have some playbooks of its own and run them. This process will also be used for updates. It is the main rpi which will update itself and the rpi with camera. It is nice to know whether or not these updates have succeeded. Here I am using an Incoming webhook from Slack to send events play following this [doc](https://docs.ansible.com/ansible/latest/collections/community/general/slack_callback.html#parameters). To get the `SLACK_WEBHOOK_URL` in use on Pyronear Slack send me a dm. 

The boot sends messages like this:

Capture d’écran 2021-04-11 à 21 13 50

File: files/ansible_on_main_rpi/README.md
Comment (chloeskt) [inline files/ansible_on_main_rpi/README.md:1]: README to explain the setup of the directory containing all Ansible files and playbooks on the main RPI.

File: files/ansible_on_main_rpi/hosts.yml
Comment (chloeskt) [inline files/ansible_on_main_rpi/hosts.yml:2]: In this hosts.yml file we will need to put the fixed IP adresses of the RPI camera

File: files/ansible_on_main_rpi/playbooks/configure_rpi_camera.yml
Comment (chloeskt) [inline files/ansible_on_main_rpi/playbooks/configure_rpi_camera.yml:1]: This is a setup playbook, it will be run when we do the setup of our fleet of rpi

File: files/ansible_on_main_rpi/templates/python.env.dist
Comment (chloeskt) [inline files/ansible_on_main_rpi/templates/python.env.dist:1]: This env file is needed so that the camera rpi know the url to do the post request in the monitor_pi.py script.

- `MAIN_RPI_IP` is the fix IP address of the master rpi
- `MAIN_RPI_WEBSERVER_PORT` is the port of the webserver

File: playbooks/add_jobs_crontab.yml
Comment (chloeskt) [inline playbooks/add_jobs_crontab.yml:10]: This job is put on a cron tab so that it is always run. In `monitor_pi.py` there is a while True loop (which also caches errors) and we put this script in a cron job launch at reboot (so that if the rpis crash, the script will be launched again automatically).

File: playbooks/install_ansible_rpi_main.yml
Comment (chloeskt) [inline playbooks/install_ansible_rpi_main.yml:2]: This playbook also us to copy the directory which will be installed on the rpi master containing all related Ansible conf and files.

File: playbooks/install_local_api_dependencies.yml
Comment (chloeskt) [inline playbooks/install_local_api_dependencies.yml:18]: NB: The build of the image on RPI takes a very long time.

File: playbooks/install_pyroengine.yml
Comment (chloeskt) [inline playbooks/install_pyroengine.yml:17]: This allows us to recover the old hash of the last commit and the new one. Only the new one is of interest I guess, it is stored and then we will be able to use the route in the API to compare it with what it should be.

File: playbooks/install_vision_dependencies.yml
Comment (chloeskt) [inline playbooks/install_vision_dependencies.yml:28]: This is necessary to install pyrovision, does not work otherwise (need strict req matching the builds previously installed).

File: templates/vars.yml.j2
Comment (chloeskt) [inline templates/vars.yml.j2:1]: Again needed to integrate slack.

File: playbooks/install_local_api_dependencies.yml
Comment (fe51) [reply playbooks/install_local_api_dependencies.yml:18]: Even with the last image from mateo hub ? 
About how long ?

File: playbooks/install_vision_dependencies.yml
Comment (fe51) [reply playbooks/install_vision_dependencies.yml:28]: Did you try to run the webserver (having vision dependencies) without those packages installed (as long as they are compiled for the docker image) ?

**Review by fe51 — APPROVED** [review]
Hello Chloé, 

Thanks a lot for this PR ! 

I have not tested on pi yet, but all the flow is clear. 

I have left a few questions. As long as we are going to iterate, and this PR brings multiple new feature, I think this is okay for merging  ! 

(I will try to test everything on pi quite soon)

**Comment by fe51** [comment]
And thanks for the guiding comments 👍

File: playbooks/install_vision_dependencies.yml
Comment (chloeskt) [reply playbooks/install_vision_dependencies.yml:28]: No ! As of when I opened this PR, the webserver had no vision dependencies. I wanted to wait until this PR was merged to work on integrating the latest enhancement of pyroengine and checking that everything is working correctly (not done in the scope of this PR)

File: main.rpi.requirements.txt
Comment (chloeskt) [reply main.rpi.requirements.txt]: You are totally right ! It will change it in an upcoming PR to `raspberrypi_main` for the hosts names.

File: playbooks/install_local_api_dependencies.yml
Comment (chloeskt) [reply playbooks/install_local_api_dependencies.yml:18]: Did not try it, as I say, I did not touch this PR since it was ready for review because I think that the integration of the changes in pyroengine since 5/6 days ago will change stuff.

**Pull request merged**
